### PR TITLE
Row contains

### PR DESCRIPTION
**Pull request by justinwoo**
add Prim.Row.Contains solved type class

depends on #3389

### CODE REVIEW

**Comment by justinwoo** [comment]
Errored out at cloning github repos as usual... but I guess we can come back to this in a bit

**Comment by justinwoo** [comment]
This was fixed with some amended commits

File: src/Language/PureScript/TypeChecker/Entailment.hs [hunk 1]
```diff
@@ -496,6 +497,22 @@ entails SolverOptions{..} constraint context hints =
             REmpty -> (True, Nothing)
             _ -> (not (null fixed), Just [ Constraint C.RowLacks [TypeLevelString sym, rest] Nothing ])
 
+    solveRowContains :: [Type] -> Maybe [TypeClassDict]
+    solveRowContains [TypeLevelString sym, r, _] =
+        pure [ TypeClassDictionaryInScope [] 0 EmptyClassInstance [] C.RowContains [TypeLevelString sym, r, result] Nothing ]
```
[the remark this answers is not in the document]
Comment (justinwoo) [reply src/Language/PureScript/TypeChecker/Entailment.hs]: I need to supply not only the rest of the row but also a result placeholder though, as `RowContains "x" s result`. Is there somewhere I can find this?

[the remark this answers is not in the document]
Comment (LiamGoodacre) [reply src/Language/PureScript/TypeChecker/Entailment.hs]: I think you can reuse the last argument: the `_` in `solveRowContains [TypeLevelString sym, r, _] =`.

[the remark this answers is not in the document]
Comment (justinwoo) [reply src/Language/PureScript/TypeChecker/Entailment.hs]: Cool, I tried it out and added a test that uses it

File: src/Language/PureScript/Environment.hs [hunk 2]
```diff
@@ -518,6 +519,15 @@ primRowClasses =
         , ("row", Just (Row kindType))
         ] [] [] [])
 
+    -- class Contains (label :: Symbol) (row :: # Type) (result :: Boolean)
```
Comment (LiamGoodacre) [inline src/Language/PureScript/Environment.hs]: These comments usually also mention the functional dependencies.

File: src/Language/PureScript/Environment.hs [hunk 3]
```diff
@@ -518,6 +519,15 @@ primRowClasses =
...
+    , (primSubName C.moduleRow "Contains", makeTypeClassData
+        [ ("label", Just kindSymbol)
+        , ("row", Just (Row kindType))
+        , ("result", Just (Row kindBoolean))
```
Comment (LiamGoodacre) [inline src/Language/PureScript/Environment.hs]: `result` should be of kind `Boolean` not `# Boolean`.

File: src/Language/PureScript/Constants.hs [hunk 4]
```diff
@@ -457,6 +460,12 @@ pattern SymbolAppend = Qualified (Just PrimSymbol) (ProperName "Append")
 pattern SymbolCons :: Qualified (ProperName 'ClassName)
 pattern SymbolCons = Qualified (Just PrimSymbol) (ProperName "Cons")
 
+pattern SymbolContains :: Qualified (ProperName 'ClassName)
+pattern SymbolContains = Qualified (Just PrimSymbol) (ProperName "Contains")
+
+pattern SymbolBreakOn :: Qualified (ProperName 'ClassName)
+pattern SymbolBreakOn = Qualified (Just PrimSymbol) (ProperName "BreakOn")
```
Comment (LiamGoodacre) [inline src/Language/PureScript/Constants.hs]: This isn't supposed to be in this PR is it?

Comment (justinwoo) [reply src/Language/PureScript/Constants.hs]: oh, rebase failure

**Comment by justinwoo** [comment]
updated to latest changes in master, requested changes has already been applied

**Comment by joelmccracken** [comment]
What is wrong with this PR?

**Comment by justinwoo** [comment]
Cleaning up my github notifications

**Pull request closed without merging**